### PR TITLE
stm32h7 add entropy generator support

### LINKED ISSUE #27373: CivetWeb Support for STM32H7 Series
**Describe the bug**
Building `samples/net/sockets/civetweb` for `nucleo_h745zi_q` produces following warning:

```
warning: MINIMAL_LIBC_MALLOC_ARENA_SIZE (defined at lib/libc/Kconfig:92) was assigned the value
'131072' but got the value ''. Check these unsatisfied dependencies: MINIMAL_LIBC_MALLOC (=n),
MINIMAL_LIBC (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE.html
and/or look up MINIMAL_LIBC_MALLOC_ARENA_SIZE in the menuconfig/guiconfig interface. The Application
Development Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of
the manual might be helpful too.
```

This warning leads to following error in example's compilation:
```
In file included from <your_path_here>/zephyrproject/zephyr/samples/net/sockets/civetweb/src/libc_extensions.c:9:
<your_path_here>/zephyrproject/zephyr/samples/net/sockets/civetweb/src/libc_extensions.h:42:5: error: expected identifier or '(' before 'int'
   42 | int ferror(FILE *stream);
```

**To Reproduce**
> west build -p auto -b "nucleo_h745zi_q_m7" "samples/net/sockets/civetweb"

**Expected behavior**
A clearly compiled and running project on mentioned board.

**Impact**
Non-compilable default sample for mentioned board

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK v0.11.4
 - Commit SHA or Version used: 5328adcff5e110559f2b179318102e8700a7e172

**Additional context**
My investigations showed, that there is explicit `fulllibc` select in `stm32h7` series and no random generator support.
P.S.: The mentioned sample project can be perfectly compiled with `nucleo_f767zi` board.

**Possible solution**
1. Add `RNG` support to stm32h7 series ( can be solved by #27374 )
2. Remove explicit `fulllibc` select for stm32h7 series ( can be solved by #27389 )

### PR DESCRIPTION
**Pull request by KozhinovAlexander**
Adds entropy generator support and removes explicit fulllibc support.

This PR needed for #27373

### CODE REVIEW

**Comment by Nukersson** [comment]
@ABOSTM @gmarull @carlescufi Please for review at least from two of you. The changes are very trivial and small.

File: soc/arm/st_stm32/stm32h7/Kconfig.series [hunk 1]
```diff
@@ -12,7 +12,6 @@ config SOC_SERIES_STM32H7X
 	select HAS_STM32CUBE
 	select CPU_HAS_ARM_MPU
 	select HAS_SWO
-	select REQUIRES_FULL_LIBC
```
Comment (gmarull) [inline soc/arm/st_stm32/stm32h7/Kconfig.series:15]: Why is this required?

Comment (Nukersson) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: 1. otherwise `minlibc` can not be used and compiling of `samples/net/sockets/civetweb` produces following warning:
```
warning: MINIMAL_LIBC_MALLOC_ARENA_SIZE (defined at lib/libc/Kconfig:92) was assigned the value
'131072' but got the value ''
```
2. `soc/arm/st_stm32f7/stm32f7/Kconfig.series` - does not contain `REQUIRES_FULL_LIBC` also

Comment (ABOSTM) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: I 'm concerned about the impact of not selecting REQUIRES_FULL_LIBC?
Could it cause regression other tests or samples ?
I would prefer to wait for @erwango feedback

**Review by ABOSTM — CHANGES_REQUESTED** [review]
@Nukersson , you enabled rng peripheral for both core.
But, as a general rule for STM32 dualcore in zephyr, to avoid race condition on peripheral registers, only 1 core is allowed to access a peripheral (with some exceptions like RCC registers, flash, gpio, ...). 
So rng should be enabled on 1 dts only.
If it is really necessary to use rng on both core, you could protect rng with hardware semaphore (HSEM), but it is a much more complex (and it is not the current way of working in zephyr):
* you can take example from PR #24862 to see how to use HSEM zephyr/stm32 API
* you MUST take the PR #26929 , which is not yet merged
  please note that PR #26929 tackled RNG Clock configuration which is common to both USB and RNG, but it doesn't tackled concurrent access to RNG registers
* I am not RNG expert, but AFAIK random number generation can take some time and thus It may be necessary to adjust HSEM timeout consequently

**Comment by Nukersson** [comment]
> @Nukersson , you enabled rng peripheral for both core.
> But, as a general rule for STM32 dualcore in zephyr, to avoid race condition on peripheral registers, only 1 core is allowed to access a peripheral (with some exceptions like RCC registers, flash, gpio, ...).
> So rng should be enabled on 1 dts only.
> If it is really necessary to use rng on both core, you could protect rng with hardware semaphore (HSEM), but it is a much more complex (and it is not the current way of working in zephyr):
> 
> * you can take example from PR #24862 to see how to use HSEM zephyr/stm32 API
> * you MUST take the PR #26929 , which is not yet merged
>   please note that PR #26929 tackled RNG Clock configuration which is common to both USB and RNG, but it doesn't tackled concurrent access to RNG registers
> * I am not RNG expert, but AFAIK random number generation can take some time and thus It may be necessary to adjust HSEM timeout consequently

Okay. For the sake of simplicity I would prefer to leave RNG support only to M7 core for the first. The change comes with next push.

File: soc/arm/st_stm32/stm32h7/Kconfig.series (hunk 1 above)
Comment (Nukersson) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: > I 'm concerned about the impact of not selecting REQUIRES_FULL_LIBC?
> Could it cause regression other tests or samples ?
> I would prefer to wait for @erwango feedback

I would say - see 2. in my reply above. And also buildkite runs properly. But as you wish - we can wait for @erwango

Comment (Nukersson) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: @ABOSTM @gmarull Plese see description of #27373 to compile `samples/net/sockets/civetweb` for one of stm32h7 boards. And you will see which impact explicit select of `REQUIRES_FULL_LIBC` has on this sample. I just followed same approach from stm32f7 series, where `REQUIRES_FULL_LIBC` does not selected explicitly (see description of #27373 for it also).

Comment (ABOSTM) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: > soc/arm/st_stm32f7/stm32f7/Kconfig.series - does not contain REQUIRES_FULL_LIBC also

That doesn't mean H7 don't require it. There may be other solutions that keep REQUIRES_FULL_LIBC and remove your errors ...
My understanding is that there are 2 issues: 
1) The warning which comes from the fact that the test defines CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE (prj.conf)  even when not needed (when REQUIRES_FULL_LIBC is defined). This could for example be solved by changing the sample.
2) The compilation error. This look to me like the test is not designed to work with the full LibC.
So now, 
- should we adapt the sample to work with both non-full and full libC? 
- exclude the board for this test (as it is not compatible with full libC)? 
- or change the driver to not use the Full LibC (like you proposed) ?
Honestly I don't know. And prefer to open the discussion.

> And you will see which impact explicit select of `REQUIRES_FULL_LIBC` has on this sample

My concern is not about this sample, but about the impact on all others samples and tests.
I want to avoid regressions.

Comment (gmarull) [reply soc/arm/st_stm32/stm32h7/Kconfig.series:15]: I think that the intent of this PR is to add RNG support, so this parameter should not be changed here. If this change is required for something else I think a new PR should be created with some context to better understand the implications.

Comment (Nukersson) [reply soc/arm/st_stm32/stm32h7/Kconfig.series]: @gmarull  @ABOSTM @erwango  The commit which is the point of this discussion were removed from this PR. Another PR #27389 for fulllibc explicit select remove were opened.

**Comment by Nukersson** [comment]
@ABOSTM @gmarull Please review new revision of this PR without deleting of `fulllibc` support (see last push).

**Review by ABOSTM — APPROVED** [review]
LGTM

**Comment by gmarull** [comment]
@Nukersson LGTM, but remove the “fixes N” from the description, otherwise the issue will get closed and this PR does not fix that issue now.

**Comment by Nukersson** [comment]
> @Nukersson LGTM, but remove the “fixes N” from the description, otherwise the issue will get closed and this PR does not fix that issue now.

Done

**Review by gmarull — APPROVED** [review]
